File: 24_swapPairs.cpp
// 24. Swap Nodes in Pairs

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head==nullptr or head->next==nullptr)
            return head;
        
        ListNode* node1 = head;
        ListNode* node2 = head->next;
        ListNode* node3 = head->next->next;
        node2->next = node1;
        
        ListNode* node4 = swapPairs(node3);
        node1->next = node4;
        
        return node2;
        
    }
};

File: 24_swapPairs.py
# 24. Swap Nodes in Pairs

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if not head or not head.next:
            return head
        
        node1 = head
        node2 = head.next
        node3 = head.next.next
        node2.next = node1
        
        node4 = self.swapPairs(node3)
        node1.next = node4
        
        return node2
